docs(service-worker): add staleWhileRevalidate strategy

ajitsinghkaler opened the pull request:
There is great workaround for implementing staleWhileRevalidate strategy in service-worker by setting strategy to freshness and timeout to 0u. Documented this in service worker config where all other strategies are documented

Fixes #20402

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
NO staleWhileRevalidate strategy in service worker


Issue Number: #20402


## What is the new behavior?
Added staleWhileRevalidate strategy in service worker

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mary-poppins commented:
You can preview b4b6f67 at https://pr37301-b4b6f67.ngbuilds.io/.

gkalpak commented on aio/content/guide/service-worker-config.md:
```diff
@@ -193,6 +193,24 @@ The Angular service worker can use either of two caching strategies for data res
+
+<div class="alert is-helpful">
+
+You can also implement a third strategy ie
+[staleWhileRevalidate](https://developers.google.com/web/fundamentals/instant-and-offline/offline-cookbook/#stale-while-revalidate) which shows cached data until new data is not
+featched from the network. To implement this strategy set strategy to 'freshness' and timeout:
+0u in cacheConfig.
```
```suggestion
You can also emulate a third strategy, [staleWhileRevalidate](https://developers.google.com/web/fundamentals/instant-and-offline/offline-cookbook/#stale-while-revalidate), which returns cached data (if available), but also fetches fresh data from the network in the background for next time.
To use this strategy set `strategy` to `freshness` and `timeout` to `0u` in `cacheConfig`.
```

gkalpak commented on aio/content/guide/service-worker-config.md:
```diff
@@ -193,6 +193,24 @@ The Angular service worker can use either of two caching strategies for data res
+featched from the network. To implement this strategy set strategy to 'freshness' and timeout:
+0u in cacheConfig.
+
+What this will essentially do is:
+
+1. Try to fetch from network first.
+2. If the network request does not complete after the 0ms timeout (means immediately),
+fallback to the cache.
+3. Once the network request completes, update the cache.
+4. If the resource does not exist in the cache, wait for the network request anyway.
```
```suggestion
This will essentially do the following:

1. Try to fetch from the network first.
2. If the network request does not complete after 0ms (i.e. immediately), fall back to the cache (ignoring cache age).
3. Once the network request completes, update the cache for future requests.
4. If the resource does not exist in the cache, wait for the network request anyway.
```

gkalpak reviewed (changes requested):
Thx, @ajitsinghkaler ✨
I made some suggestions wrt wording (and some extra details). Otherwise, I think this is a nice addition 👌

ajitsinghkaler committed:
docs(service-worker): add staleWhileRevalidate strategy

There is great workaround for implementing staleWhileRevalidate strategy in service-worker by setting strategy to freshness and timeout to 0u. Documented this in service worker config where all other strategies are documented

Fixes #20402

mary-poppins commented:
You can preview c8fef55 at https://pr37301-c8fef55.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made the changes. Please have a look.

gkalpak reviewed (approved):
LGTM :sparkles:
Thx, @ajitsinghkaler :+1:
